### PR TITLE
ENH: Support complex-valued images and kernels for many ndimage filters

### PR DESCRIPTION
**Pull request by grlee77**
#### Reference issue
None

#### What does this implement/fix?
This PR allows the functions within `scipy.ndimage.filters` that are based on correlation or convolution to operate on complex-valued `input` and/or with complex-valued `weights`. Examples are:

`convolve`, `correlate`, `gaussian_filter`, `laplace`, `prewitt`, `sobel`, `uniform_filter`, etc.

Complex values remain unsupported by `generic_filter` and the various rank filters.

#### Additional information
The `weights` in `ndimage.correlate` are conjugated. This definition of correlation is consistent with `numpy.correlate`, where the second argument is conjugated. It is also consistent with the definition used by `xcorr` and `xcorr2` in Matlab's signal processing toolbox.

The general approach is to decompose the convolution of two complex-valued components into a combination of four real-valued convolutions. These are the equivalent of the four terms on the right hand side of the complex scalar multiplication shown below:

**(a + jb) * (g + jh) = ag + jbg + jah - bh**

If either the `input` or `weights` are real then only two real-valued convolutions need to be combined.

This approach is MUCH simpler than modifying the underlying ndimage C code which does not currently have complex-value support anywhere. Adding it there would require much more work, not only in `ni_filters.c` but also in `ni_support.c`, because the line buffers used by ndimage are currently always allocated in (real) double precision.

### CODE REVIEW

**Comment by grlee77** [comment]
I have rebased to resolve the conflicts following the ndimage test reorganization in #12746

**Review by tylerjereddy — COMMENTED** [review]
I read through all the code changes just now. Looks solid/safe from the standpoint that you're adding support for a new feature and have not changed any old unit tests apart from expanding the scope of the types they test.

Two of the three Travis CI failures are real but should be easy to fix: lint/formatting failure and a refguide check docstring failure.

One more comment--is there someone you can ping or an issue/comment you can point to confirming this is something we want? Just asking because I'm not a domain expert.

**Comment by tylerjereddy** [comment]
The desire for this feature may also have been in a proposal/small grant you were working on? If so, that makes sense--just wanted to confirm.

**Comment by grlee77** [comment]
Thanks, Travis was taking forever to start, so I hadn't gotten back to check on it.

A similar PR for the interpolation module (#12812) was mentioned in the small grant proposal, but more as an optional component and not the main deliverable. I don't think this one specific to the filtering module was mentioned there.

In general, I have sometimes wanted this for filtering complex-valued images with real-valued filters. While it is not difficult to manually do this on the separate components it is a little annoying, particularly if the output argument is also used. I think that is not too uncommon a desire. See for instance:
https://dsp.stackexchange.com/questions/31523/smoothing-complex-data-by-convolution

I am less sure about the need to support complex-valued weights. A google search did reveal quite a few publications involving complex weights, but these were mostly in the context of deep learning.

There have also been requests to other imaging libraries for complex support, although the following two appear to be in the context of interpolation rather than filtering:
https://github.com/ANTsX/ANTs/issues/565
https://discourse.itk.org/t/support-for-complex-pixel-type-in-simpleitk-is-lacking/1942/5

File: scipy/ndimage/tests/test_filters.py [hunk 1]
```diff
@@ -12,15 +13,102 @@
...
+            func(array.real, kernel.real, output=real_dtype) -
+            func(array.imag, kernel.imag, output=real_dtype) +
+            1j * func(array.imag, kernel.real, output=real_dtype) +
+            1j * func(array.real, kernel.imag, output=real_dtype)
+        )
+    elif complex_array:
+        output = (
+            func(array.real, kernel, output=real_dtype) +
+            1j * func(array.imag, kernel, output=real_dtype)
+        )
```
Comment (larsoner) [inline scipy/ndimage/tests/test_filters.py:49]: Can we get a little bit of memory efficiency and no need for `+` via something like:
```
output = np.empty(array.shape, dtype=np.complex128).view(np.float64)
func(array.real, kernel, output=output[..., ::2])
func(array.imag, kernel, output=output[..., 1::2])
```

**Review by larsoner — APPROVED** [review]
Code LGTM other than one potential memory-saving idea

File: scipy/ndimage/tests/test_filters.py (hunk 1 above)
Comment (grlee77) [reply scipy/ndimage/tests/test_filters.py:49]: I think so, but I left it that way by intention to try and make this test reference implmentation unambiguous. This version is only use to validate the true implementation in `scipy.ndimage.filters._complex_via_real_components`, so it only ever receives tiny images from the test cases. 

Currently it takes around 6 seconds on my machine to run the full ndimage test suite, so I don't think it is too critical to reduce the test times further.

I am interested in having the user-facing implementation within the `_complex_via_real_components` function within `filters.py` be efficient as possible. Within that version, I did do try to do this type of thing (providing `output=output.real`, etc).

Comment (larsoner) [reply scipy/ndimage/tests/test_filters.py:49]: Whoops, I didn't even notice this was in `test_*` when I was reading this! Yes I agree it's preferable in that case to leave it as the current, clearer implementation

**Comment by tylerjereddy** [comment]
Ok, there are basically two core developer approvals on this PR now. Travis CI is hosed/backlogged for > 7 hours because of their recent policy changes presumably. Everything else in CI is green.

I'm going to go ahead and trust that the linting/refguide stuff normally checked by Travis was solved in the new commits above and merge this so we can keep moving forward. Thanks @grlee77 and @larsoner

**Pull request merged**